util/lock.go

```diff
@@ -0,0 +1,170 @@
+package util
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"github.com/timescale/prometheus-postgresql-adapter/log"
+	"strconv"
+	"sync"
+	"time"
+)
+
+const (
+	connCheckInterval = time.Millisecond * 250
+)
+
+// PgAdvisoryLock is implementation of leader election protocol based on PostgreSQL advisory locks. All adapters withing a HA group are trying
+// to obtain an advisory lock for particular group. The one who holds the lock can write to the database. Due to the fact
+// that Prometheus HA setup provides no consistency guarantees this implementation is best effort in regards
+// to metrics that is written (some duplicates or data loss are possible during failover)
+type PgAdvisoryLock struct {
+	conn     *sql.Conn
+	mutex    sync.RWMutex
```

Comment:
Ah, OK, this was my confusion with the test as well. I saw the test that you wrote, but that wasn't what I had in mind, because I thought you would want to test when clients are in completely different processes. But I guess that's not how the adapter is executed. I'm still not seeing how a single `PgAdvisoryLock` is shared across (requests?) but maybe that is code that is outside of this PR. I'll ask you offline.

Reply:
IMHO having separate processes would not bring any benefit to testing but rather increase test complexity/fragility a lot. `PgAdvisoryLock` is being shared among go routines which are created on each request: https://golang.org/pkg/net/http/#Server.Serve